lib/invoker/power/setup/distro/ubuntu.rb

```diff
@@ -4,6 +4,42 @@ module Invoker
   module Power
     module Distro
       class Ubuntu < Debian
+        def using_systemd_resolved?
+          return @_using_systemd_resolved if defined?(@_using_systemd_resolved)
+          @_using_systemd_resolved = system("systemctl is-active --quiet systemd-resolved")
+        end
+
+        def install_required_software
+          if using_systemd_resolved?
+            # Don't install dnsmasq if Ubuntu version uses systemd-resolved for DNS because they conflict
+            system("apt-get --assume-yes install socat")
+          else
+            super
+          end
+        end
+
+        def install_packages
+          using_systemd_resolved? ? "socat" : super
+        end
+
+        def install_other
+          using_systemd_resolved? ? nil : super
+        end
+
+        def resolver_file
+          using_systemd_resolved? ? nil : super
+        end
```

Comment:
If I understand this correctly, we don't need socat when systemd-resolvd is being used? So why do we even install socat.

Reply:
It would be nice to read list of packages from same `install_packages` method here and for printing to user rather than duplicating the information in two places. But that duplication has always been there.

Reply:
@gnufied as far as I'm aware, socat performs port proxying (eg: domain.localhost:80 -> 127.0.0.1:9000). I tried on Ubuntu 18 without it and it required accessing domain.localhost:9000 but I'm open to having someone more experienced test